<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • chromVAR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../jquery.sticky-kit.min.js"></script><script src="../../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../../index.html">chromVAR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/Introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../../articles/Articles/Counts.html">Counts</a>
    </li>
    <li>
      <a href="../../articles/Articles/Annotations.html">Annotations</a>
    </li>
    <li>
      <a href="../../articles/Articles/Deviations.html">Deviations</a>
    </li>
    <li>
      <a href="../../articles/Articles/Applications.html">Applications</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/GreenleafLab/chromVAR">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1></h1>
            
          </div>

    
    
<div class="contents">
<div id="annotations" class="section level1">
<h1 class="hasAnchor">
<a href="#annotations" class="anchor"></a>Annotations</h1>
<p>The main <code>computeDeviations</code> function from chromVAR requires an object storing what peaks overlap what motifs or other annotations. The package includes functions for creating such an object from a set of motifs or kmers, or for converting an existing matrix, data.frame, list, GenomicRangesList, or list of bed files of annotations into the appropriate object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(chromVAR)
<span class="kw">library</span>(motifmatchr)
<span class="kw">library</span>(SummarizedExperiment)
<span class="kw">library</span>(Matrix)
<span class="kw">library</span>(BiocParallel)
<span class="kw">register</span>(<span class="kw">MulticoreParam</span>(<span class="dv">8</span>))</code></pre></div>
<div id="motifs-or-kmers" class="section level2">
<h2 class="hasAnchor">
<a href="#motifs-or-kmers" class="anchor"></a>Motifs or kmers</h2>
<p>The most common type of annotations to use for chromVAR are motifs. chromVAR has a function to make it easy to read in motifs from the JASPAR database:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">jaspar_motifs &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/getJasparMotifs.html">getJasparMotifs</a></span>() <span class="co"># default species is human</span></code></pre></div>
<div id="package-with-motifs" class="section level3">
<h3 class="hasAnchor">
<a href="#package-with-motifs" class="anchor"></a>Package with motifs</h3>
<p>A companion package, chromVARmotifs, includes pwms from a couple different sources that can also be used with motifmatchr and chromVAR.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Not Run</span>
devtools::<span class="kw">install_github</span>(<span class="st">"GreenleafLab/chromVARmotifs"</span>)

<span class="kw">data</span>(<span class="st">"human_pwms_v1"</span>) <span class="co"># human collection</span>

<span class="kw">data</span>(<span class="st">"mouse_pwms_v1"</span>) <span class="co"># mouse collection</span>

<span class="kw">data</span>(<span class="st">"homer_pwms"</span>)

<span class="kw">data</span>(<span class="st">"encode_pwms"</span>)</code></pre></div>
<p>If using your own motifs, they need to be formatted as either a PWMatrixList or a PFMatrixList from the TFBSTools package. See documentation from TFBStools for those two objects.</p>
</div>
<div id="finding-motif-matches" class="section level3">
<h3 class="hasAnchor">
<a href="#finding-motif-matches" class="anchor"></a>Finding motif matches</h3>
<p>The package motifmatchr can be used to find motifs within peaks. The <code>matchMotifs</code> method takes as inputs the motif lists described above, or your own list of motifs, and returns a (Ranged)SummarizedExperiment with a matrix indicating what peaks (rows) contain what motifs (columns).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(BSgenome.Hsapiens.UCSC.hg19)</code></pre></div>
<pre><code>## Loading required package: BSgenome</code></pre>
<pre><code>## Loading required package: Biostrings</code></pre>
<pre><code>## Loading required package: XVector</code></pre>
<pre><code>## 
## Attaching package: 'Biostrings'</code></pre>
<pre><code>## The following object is masked from 'package:DelayedArray':
## 
##     type</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     strsplit</code></pre>
<pre><code>## Loading required package: rtracklayer</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># First get filtered counts</span>
<span class="kw">data</span>(example_counts, <span class="dt">package =</span> <span class="st">"chromVAR"</span>)
example_counts &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/addGCBias.html">addGCBias</a></span>(example_counts, <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19)
counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/filterSamples.html">filterSamples</a></span>(example_counts, <span class="dt">min_depth =</span> <span class="dv">1500</span>, <span class="dt">min_in_peaks =</span> <span class="fl">0.15</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)
counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/filterPeaks.html">filterPeaks</a></span>(counts_filtered)

<span class="co"># get motif matches</span>
motif_ix &lt;-<span class="st"> </span><span class="kw">matchMotifs</span>(jaspar_motifs, counts_filtered, <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19)</code></pre></div>
<p><code>matchMotifs</code> can accept as the second argument either a RangedSummarizedExperiment, GenomicRanges, DNAStringSet, or character vector. If the argument is not a sequence object, then a genome argument is also required, which should be a BSgenome object – the default is BSgenome.Hsapiens.UCSC.hg19. For more information on motifmatchr see the vignette from that package.</p>
</div>
<div id="using-kmers" class="section level3">
<h3 class="hasAnchor">
<a href="#using-kmers" class="anchor"></a>Using kmers</h3>
<p>The <code>matchKmers</code> function in chromVAR can be used to make an annotation matrix for all kmers of a given length:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">kmer_ix &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/matchKmers.html">matchKmers</a></span>(<span class="dv">6</span>, counts_filtered, <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19)</code></pre></div>
<p>Alternatively, if you only want an annotation matrix for a set of pre-determined kmers, you can pass <code>matchKmers</code> a list of those kmers:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_kmers_ix &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/matchKmers.html">matchKmers</a></span>(<span class="kw">c</span>(<span class="st">"AGATAA"</span>,<span class="st">"GATAAG"</span>), counts_filtered, 
                          <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19)</code></pre></div>
</div>
<div id="cis-groups" class="section level3">
<h3 class="hasAnchor">
<a href="#cis-groups" class="anchor"></a>Cis groups</h3>
<p>Instead of getting groups of peaks based on shared motif or kmer, chromVAR can also use peak groups defined based on location. The fuction getCisGroups will annotate peaks based on chromosomal location. With default parameters, the function will take the first 25 peaks in a chromosome and consider those a group, then move 10 peaks down and group the 25 peaks together, and so on, to create overlapping sets of 25 peaks. The group size and step size can both be adjusted through the paremeters grpsize and stepsize.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cis_ix &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/getCisGroups.html">getCisGroups</a></span>(counts_filtered, <span class="dt">grpsize =</span> <span class="dv">25</span>, <span class="dt">stepsize =</span> <span class="dv">10</span>) </code></pre></div>
</div>
</div>
<div id="alternative-annotations" class="section level2">
<h2 class="hasAnchor">
<a href="#alternative-annotations" class="anchor"></a>Alternative annotations</h2>
<div id="reading-in-annotations-from-bed-files" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-in-annotations-from-bed-files" class="anchor"></a>Reading in annotations from bed files</h3>
<p>If your have a set of genomic annotations (e.g. genomic motif matches, GWAS hits) in bed files that you want to use, then the function <code>getAnnotations</code> can be used to read those annotations into the appropriate SummarizedExperiment with matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Not evaluated</span>
my_annotation_files &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">'anno1.bed'</span>,<span class="st">'anno2.bed'</span>) <span class="co"># Fake files</span>
anno_ix &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/getAnnotations.html">getAnnotations</a></span>(my_annotation_files, 
                          <span class="dt">rowRanges =</span> <span class="kw">rowRanges</span>(counts_filtered))</code></pre></div>
</div>
<div id="reading-in-annotations-from-single-bed-file-with-column-indicating-specific-annotation-" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-in-annotations-from-single-bed-file-with-column-indicating-specific-annotation-" class="anchor"></a>Reading in annotations from single bed file with column indicating specific annotation.</h3>
<p>Use the <code>getAnnotations</code> with a single file and the column argument to get an annotation matrix based on the groups indicated in that column (1-based indexing).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Not evaluated</span>
my_annotation_file &lt;-<span class="st"> "anno.bed"</span> <span class="co"># Fake file</span>
anno_ix &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/getAnnotations.html">getAnnotations</a></span>(my_annotation_files, 
                           <span class="dt">rowRanges =</span> <span class="kw">rowRanges</span>(counts_filtered), <span class="dt">column =</span> <span class="dv">5</span>)</code></pre></div>
</div>
<div id="other-annotation-formats" class="section level3">
<h3 class="hasAnchor">
<a href="#other-annotation-formats" class="anchor"></a>Other annotation formats</h3>
<p>If you have already read in your annotations into R, you can also use the <code>getAnnotations</code> function to get them into the right format.</p>
<div id="annotations-genomicrangeslist" class="section level4">
<h4 class="hasAnchor">
<a href="#annotations-genomicrangeslist" class="anchor"></a>Annotations GenomicRangesList:</h4>
<p>If your annotations are stored as a GenomicRangesList, with each element of the list a GenomicRanges corresponding to the locations of the annotation, then the <code>getAnnotations</code> method can determine what peaks overlap with what annotations and create the appropriate matrix stored within the <code>annotationMatches</code> assay of a SummarizedExperiment.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GenomicRanges)
my_annotation_granges &lt;-<span class="st"> </span><span class="kw">GRangesList</span>(<span class="kw">GRanges</span>(<span class="st">"chr1"</span>, 
                                             <span class="dt">ranges =</span> <span class="kw">IRanges</span>(<span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">566763</span>,<span class="dv">805090</span>), 
                                                              <span class="dt">width =</span> <span class="dv">8</span>)),
                                        <span class="kw">GRanges</span>(<span class="st">"chr1"</span>, 
                                                <span class="dt">ranges =</span> <span class="kw">IRanges</span>(<span class="dt">start =</span> <span class="kw">c</span>(<span class="dv">566792</span>,<span class="dv">895798</span>), 
                                                                 <span class="dt">width =</span> <span class="dv">8</span>)))
anno_ix &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/getAnnotations.html">getAnnotations</a></span>(my_annotation_granges, 
                          <span class="dt">rowRanges =</span> <span class="kw">rowRanges</span>(counts_filtered))</code></pre></div>
</div>
<div id="annotations-as-a-matrix-matrix-data-frame-or-dataframe" class="section level4">
<h4 class="hasAnchor">
<a href="#annotations-as-a-matrix-matrix-data-frame-or-dataframe" class="anchor"></a>Annotations as a matrix, Matrix, data.frame, or DataFrame</h4>
<p>If your annotations already exist as a matrix-like object indicating what peak contains what annotation, then the <code>getAnnotations</code> method is simply a wrapper around SummarizedExperiment that ensures that the annotation gets stored in the <code>annotationMatches</code> assay.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_annotation_df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">first100 =</span> <span class="kw">c</span>(<span class="kw">rep</span>(<span class="ot">TRUE</span>, <span class="dv">100</span>),<span class="kw">rep</span>(<span class="ot">FALSE</span>,<span class="kw">nrow</span>(counts_filtered) -<span class="st"> </span><span class="dv">100</span>)),
                             <span class="dt">last100 =</span> <span class="kw">c</span>(<span class="kw">rep</span>(<span class="ot">FALSE</span>,<span class="kw">nrow</span>(counts_filtered) -<span class="st"> </span><span class="dv">100</span>),<span class="kw">rep</span>(<span class="ot">TRUE</span>, <span class="dv">100</span>)))

anno_ix &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/getAnnotations.html">getAnnotations</a></span>(my_annotation_df, 
                          <span class="dt">rowRanges =</span> <span class="kw">rowRanges</span>(counts_filtered))</code></pre></div>
</div>
<div id="annotations-as-a-list-of-peak-indices" class="section level4">
<h4 class="hasAnchor">
<a href="#annotations-as-a-list-of-peak-indices" class="anchor"></a>Annotations as a list of peak indices</h4>
<p>If your annotations are stored as a list of peak indices, with each element of the list a vector of which peaks contain the annotation, then the <code>getAnnotations</code> function can be used to convert that list into a boolean matrix of peak and annotation overlaps stored within the <code>annotationMatches</code> assay of a SummarizedExperiment.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my_annotation_list &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">first100 =</span> <span class="kw">rep</span>(<span class="dv">1</span>, <span class="dv">100</span>), <span class="dt">last100 =</span> <span class="kw">c</span>(<span class="kw">rep</span>(<span class="kw">nrow</span>(counts_filtered) -<span class="st"> </span><span class="dv">100</span>, <span class="kw">nrow</span>(counts_filtered))))
anno_ix &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/getAnnotations.html">getAnnotations</a></span>(my_annotation_list, 
                          <span class="dt">rowRanges =</span> <span class="kw">rowRanges</span>(counts_filtered))</code></pre></div>
<p>Alternatively, the list of peak indices can also be used as input directly into <code>computeDeviations</code>.</p>
</div>
</div>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">Sys.Date</span>()</code></pre></div>
<pre><code>## [1] "2017-10-05"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.4.1 (2017-06-30)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04.3 LTS
## 
## Matrix products: default
## BLAS: /usr/lib/libblas/libblas.so.3.6.0
## LAPACK: /usr/lib/lapack/liblapack.so.3.6.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] BSgenome.Hsapiens.UCSC.hg19_1.4.0 BSgenome_1.44.2                  
##  [3] rtracklayer_1.36.4                Biostrings_2.44.2                
##  [5] XVector_0.16.0                    BiocParallel_1.10.1              
##  [7] Matrix_1.2-11                     SummarizedExperiment_1.6.4       
##  [9] DelayedArray_0.2.7                matrixStats_0.52.2               
## [11] Biobase_2.36.2                    GenomicRanges_1.28.5             
## [13] GenomeInfoDb_1.12.2               IRanges_2.10.3                   
## [15] S4Vectors_0.14.4                  BiocGenerics_0.22.0              
## [17] motifmatchr_0.99.8                chromVAR_0.99.0                  
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                DirichletMultinomial_1.18.0
##  [3] TFBSTools_1.14.2            bit64_0.9-7                
##  [5] httr_1.3.1                  rprojroot_1.2              
##  [7] tools_3.4.1                 backports_1.1.0            
##  [9] R6_2.2.2                    DT_0.2                     
## [11] seqLogo_1.42.0              DBI_0.7                    
## [13] lazyeval_0.2.0              colorspace_1.3-2           
## [15] bit_1.1-12                  compiler_3.4.1             
## [17] plotly_4.7.1                caTools_1.17.1             
## [19] scales_0.5.0                readr_1.1.1                
## [21] stringr_1.2.0               digest_0.6.12              
## [23] Rsamtools_1.28.0            rmarkdown_1.6              
## [25] R.utils_2.5.0               pkgconfig_2.0.1            
## [27] htmltools_0.3.6             htmlwidgets_0.9            
## [29] rlang_0.1.2                 RSQLite_2.0                
## [31] VGAM_1.0-4                  shiny_1.0.3                
## [33] bindr_0.1                   jsonlite_1.5               
## [35] gtools_3.5.0                dplyr_0.7.2.9000           
## [37] R.oo_1.21.0                 RCurl_1.95-4.8             
## [39] magrittr_1.5                GO.db_3.4.1                
## [41] GenomeInfoDbData_0.99.0     Rcpp_0.12.12               
## [43] munsell_0.4.3               R.methodsS3_1.7.1          
## [45] stringi_1.1.5               yaml_2.1.14                
## [47] zlibbioc_1.22.0             plyr_1.8.4                 
## [49] grid_3.4.1                  blob_1.1.0                 
## [51] miniUI_0.1.1                CNEr_1.12.1                
## [53] lattice_0.20-35             splines_3.4.1              
## [55] annotate_1.54.0             hms_0.3                    
## [57] KEGGREST_1.16.1             knitr_1.16                 
## [59] JASPAR2016_1.4.0            reshape2_1.4.2             
## [61] TFMPvalue_0.0.6             XML_3.98-1.9               
## [63] glue_1.1.1                  evaluate_0.10.1            
## [65] data.table_1.10.4           png_0.1-7                  
## [67] httpuv_1.3.5                tidyr_0.6.3                
## [69] purrr_0.2.3                 gtable_0.2.0               
## [71] poweRlaw_0.70.1             assertthat_0.2.0           
## [73] ggplot2_2.2.1               mime_0.5                   
## [75] xtable_1.8-2                viridisLite_0.2.0          
## [77] tibble_1.3.4                GenomicAlignments_1.12.2   
## [79] AnnotationDbi_1.38.2        memoise_1.1.0              
## [81] bindrcpp_0.2</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#annotations">Annotations</a><ul class="nav nav-pills nav-stacked">
<li><a href="#motifs-or-kmers">Motifs or kmers</a></li>
      <li><a href="#alternative-annotations">Alternative annotations</a></li>
      <li><a href="#session-info">Session Info</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Alicia Schep.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
