<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • chromVAR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../jquery.sticky-kit.min.js"></script><script src="../../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../../index.html">chromVAR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/Introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../../articles/Articles/Counts.html">Counts</a>
    </li>
    <li>
      <a href="../../articles/Articles/Annotations.html">Annotations</a>
    </li>
    <li>
      <a href="../../articles/Articles/Deviations.html">Deviations</a>
    </li>
    <li>
      <a href="../../articles/Articles/Applications.html">Applications</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/GreenleafLab/chromVAR">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1></h1>
            
          </div>

    
    
<div class="contents">
<div id="applications" class="section level1">
<h1 class="hasAnchor">
<a href="#applications" class="anchor"></a>Applications</h1>
<p>This vignettes covers some of the various applications of chromVAR’s bias corrected deviations and Z-scores for motifs or other annotations. First, we will compute those deviations:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(chromVAR)
<span class="kw">library</span>(motifmatchr)
<span class="kw">library</span>(SummarizedExperiment)
<span class="kw">library</span>(Matrix)
<span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(BiocParallel)
<span class="kw">library</span>(BSgenome.Hsapiens.UCSC.hg19)
<span class="kw">register</span>(<span class="kw">MulticoreParam</span>(<span class="dv">8</span>))
<span class="kw">data</span>(example_counts, <span class="dt">package =</span> <span class="st">"chromVAR"</span>)
<span class="kw">set.seed</span>(<span class="dv">2017</span>)
example_counts &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/addGCBias.html">addGCBias</a></span>(example_counts, <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19)
counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/filterSamples.html">filterSamples</a></span>(example_counts, <span class="dt">min_depth =</span> <span class="dv">1500</span>, 
                                 <span class="dt">min_in_peaks =</span> <span class="fl">0.15</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)
counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/filterPeaks.html">filterPeaks</a></span>(counts_filtered)
motifs &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/getJasparMotifs.html">getJasparMotifs</a></span>()
motif_ix &lt;-<span class="st"> </span><span class="kw">matchMotifs</span>(motifs, counts_filtered, <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19)
dev &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/computeDeviations.html">computeDeviations</a></span>(<span class="dt">object =</span> counts_filtered, 
                                 <span class="dt">annotations =</span> motif_ix)</code></pre></div>
<div id="variability" class="section level2">
<h2 class="hasAnchor">
<a href="#variability" class="anchor"></a>Variability</h2>
<p>The first application is simply to compute the variability of each motif or annotation across the cells or samples of interest. The function <code>plotVariability</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">variability &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/computeVariability.html">computeVariability</a></span>(dev)
<span class="kw"><a href="../../reference/plotVariability.html">plotVariability</a></span>(variability, <span class="dt">use_plotly =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p><img src="Applications_files/figure-html/unnamed-chunk-2-1.png" width="672"></p>
</div>
<div id="clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering" class="anchor"></a>Clustering</h2>
<p>We can also use the bias corrected deviations to cluster the samples. The function <code>getSampleCorrelation</code> first removes highly correlated annotations and low variability annotations and then computes the correlation between the cells for the remaining annotations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sample_cor &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/getSampleCorrelation.html">getSampleCorrelation</a></span>(dev)

<span class="kw">library</span>(pheatmap)
<span class="kw">pheatmap</span>(<span class="kw">as.dist</span>(sample_cor), 
         <span class="dt">annotation_row =</span> <span class="kw">colData</span>(dev), 
         <span class="dt">clustering_distance_rows =</span> <span class="kw">as.dist</span>(<span class="dv">1</span>-sample_cor), 
         <span class="dt">clustering_distance_cols =</span> <span class="kw">as.dist</span>(<span class="dv">1</span>-sample_cor))</code></pre></div>
<p><img src="Applications_files/figure-html/unnamed-chunk-3-1.png" width="480"></p>
</div>
<div id="cell-sample-similarity" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-sample-similarity" class="anchor"></a>Cell / sample similarity</h2>
<p>We can also use tSNE for looking at cell similarity. The function <code>deviationsTsne</code> performs tsne and results a data.frame with the results. If running in an interactive session, shiny can be set to TRUE to load up a shiny gadget for exploring parameters. In general, with 100’s of cells a perplexity of around 30-50 might make sense. In this example, we have few cells so we are bounded in how high the perplexity can be set– we will use 10 here.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tsne_results &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/deviationsTsne.html">deviationsTsne</a></span>(dev, <span class="dt">threshold =</span> <span class="fl">1.5</span>, <span class="dt">perplexity =</span> <span class="dv">10</span>, 
                               <span class="dt">shiny =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>To plot the results, <code>plotDeviationsTsne</code> can be used. If running in an interactive session or an interactive Rmarkdown document, shiny can be set to TRUE to generate a shiny widget. Here we will show static results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tsne_plots &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/plotDeviationsTsne.html">plotDeviationsTsne</a></span>(dev, tsne_results, <span class="dt">annotation =</span> <span class="st">"TEAD3"</span>, 
                                   <span class="dt">sample_column =</span> <span class="st">"Cell_Type"</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)
tsne_plots[[<span class="dv">1</span>]]</code></pre></div>
<p><img src="Applications_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tsne_plots[[<span class="dv">2</span>]]</code></pre></div>
<p><img src="Applications_files/figure-html/unnamed-chunk-5-2.png" width="672"></p>
</div>
<div id="differential-accessibility-and-variability" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-accessibility-and-variability" class="anchor"></a>Differential accessibility and variability</h2>
<p>The <code>differentialDeviations</code> function determines whether there is a significant difference between the bias corrected deviations for a given annotation between different groups. The groups can be specified by giving the column name of a column in the colData of the dev object or a vector of group assignments.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">diff_acc &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/differentialDeviations.html">differentialDeviations</a></span>(dev, <span class="st">"Cell_Type"</span>)

<span class="kw">head</span>(diff_acc)</code></pre></div>
<pre><code>##        p_value p_value_adjusted
## 1 6.667785e-02     1.175006e-01
## 2 1.802723e-02     4.166773e-02
## 3 3.825026e-03     1.077708e-02
## 4 1.562799e-12     6.032403e-11
## 5 5.858065e-01     6.709831e-01
## 6 1.686300e-04     7.396724e-04</code></pre>
<p>The <code>differentialVariability</code> function determines whether there is a significant difference between the variability of any of the annotations between different groups.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">diff_var &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/differentialVariability.html">differentialVariability</a></span>(dev, <span class="st">"Cell_Type"</span>)

<span class="kw">head</span>(diff_var)</code></pre></div>
<pre><code>##     p_value p_value_adjusted
## 1 0.4330854        0.9135025
## 2 0.7011659        0.9652404
## 3 0.1308742        0.6512322
## 4 0.4142727        0.9007595
## 5 0.4509686        0.9141348
## 6 0.9073926        0.9652404</code></pre>
</div>
<div id="motif-kmer-similarity" class="section level2">
<h2 class="hasAnchor">
<a href="#motif-kmer-similarity" class="anchor"></a>Motif / kmer similarity</h2>
<p>We can also perform tsne for motif similarity rather than cell similarity, by specifying <code>what = "annotations"</code> to the <code>deviationsTsne</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">inv_tsne_results &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/deviationsTsne.html">deviationsTsne</a></span>(dev, <span class="dt">threshold =</span> <span class="fl">1.5</span>, <span class="dt">perplexity =</span> <span class="dv">8</span>, 
                                    <span class="dt">what =</span> <span class="st">"annotations"</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)

<span class="kw">ggplot</span>(inv_tsne_results, <span class="kw">aes</span>(<span class="dt">x =</span> Dim1, <span class="dt">y =</span> Dim2)) +<span class="st"> </span><span class="kw">geom_point</span>() +<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../../reference/chromVAR_theme.html">chromVAR_theme</a></span>()</code></pre></div>
<p><img src="Applications_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
</div>
<div id="kmers-and-sequence-specificity-of-variation" class="section level2">
<h2 class="hasAnchor">
<a href="#kmers-and-sequence-specificity-of-variation" class="anchor"></a>Kmers and sequence specificity of variation</h2>
<p>By using kmers as the annnotation, we can use the kmers to identify the precise nucleotides necessary for chromatin accessibility variability. The <code>deviationsCovariability</code> function returns a normalized covariance between the bias corrected deviations of any pair of annotations. We will use 6mers here in the interest of computational time, but in general 7mers yield higher variability and are better starting points for assembling de novo motifs (see next section).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">kmer_ix &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/matchKmers.html">matchKmers</a></span>(<span class="dv">6</span>, counts_filtered, <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19)
kmer_dev &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/computeDeviations.html">computeDeviations</a></span>(counts_filtered, kmer_ix)

kmer_cov &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/deviationsCovariability.html">deviationsCovariability</a></span>(kmer_dev)

<span class="kw"><a href="../../reference/plotKmerMismatch.html">plotKmerMismatch</a></span>(<span class="st">"CATTCC"</span>,kmer_cov)</code></pre></div>
<p><img src="Applications_files/figure-html/unnamed-chunk-9-1.png" width="672"></p>
</div>
<div id="de-novo-kmer-assembly" class="section level2">
<h2 class="hasAnchor">
<a href="#de-novo-kmer-assembly" class="anchor"></a>De novo kmer assembly</h2>
<p>We can use the <code>assembleKmers</code> function to build de novo motifs using the kmer deviation results. The function goes iteratively through the most variable kmers and uses the normalized covariance of the bias corrected deviations for closely related kmers to weight each nucleotide in the motif.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">de_novos &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/assembleKmers.html">assembleKmers</a></span>(kmer_dev, <span class="dt">progress =</span> <span class="ot">FALSE</span>) <span class="co">#no progress bar</span>
de_novos</code></pre></div>
<pre><code>## PWMatrixList of length 23
## names(23): denovo_1 denovo_2 denovo_3 ... denovo_21 denovo_22 denovo_23</code></pre>
<p>We can use the <code>pwmDistance</code> function to see how close our de novo motifs match known motifs. <code>pwmDistance</code> returns a list with three matrices- ‘dist’ has the distance between each pair of motifs, ‘strand’ the strand of the motif for the match, and ‘offset’ the offset between the motifs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dist_to_known &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/pwmDistance.html">pwmDistance</a></span>(de_novos, motifs)

closest_match1 &lt;-<span class="st"> </span><span class="kw">which.min</span>(dist_to_known$dist[<span class="dv">1</span>,])
dist_to_known$strand[<span class="dv">1</span>,closest_match1]</code></pre></div>
<pre><code>## [1] "-"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggmotif) <span class="co"># Package on github at AliciaSchep/ggmotif. Can use seqLogo alternatively</span>
<span class="kw">library</span>(TFBSTools)</code></pre></div>
<pre><code>## 
## Attaching package: 'TFBSTools'</code></pre>
<pre><code>## The following object is masked from 'package:Matrix':
## 
##     Matrix</code></pre>
<pre><code>## The following object is masked from 'package:DelayedArray':
## 
##     matrixClass</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggmotif_plot</span>(<span class="kw">toPWM</span>(<span class="kw">reverseComplement</span>(motifs[[closest_match1]]),<span class="dt">type =</span> <span class="st">"prob"</span>))</code></pre></div>
<p><img src="Applications_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggmotif_plot</span>(de_novos[[<span class="dv">1</span>]])</code></pre></div>
<p><img src="Applications_files/figure-html/unnamed-chunk-11-2.png" width="672"></p>
<p>In this case, it looks like the first five base pairs of the de novo motif match well with the last five of the known motif. To get a better sense of whether these might represent the same motif, one can compute the deviations for the de novo motif and see if they are correlated with the known.</p>
</div>
<div id="synergy-between-motifs" class="section level2">
<h2 class="hasAnchor">
<a href="#synergy-between-motifs" class="anchor"></a>Synergy between motifs</h2>
<p>chromVAR includes a function for computing the “synergy” between pairs of annotations/motifs, where synergy is defined as the excess variability of chromatin accessibility for peaks sharing both motifs compared to a random sub-sample of the same size of peaks with one motif (the one with greater variability).</p>
</div>
<div id="function-for-computing-synergy" class="section level2">
<h2 class="hasAnchor">
<a href="#function-for-computing-synergy" class="anchor"></a>Function for computing synergy</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(example_ix, <span class="dt">package =</span> <span class="st">"chromVAR"</span>)</code></pre></div>
<pre><code>## Warning in data(example_ix, package = "chromVAR"): data set 'example_ix'
## not found</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/chromVAR_deprecated.html">get_annotation_synergy</a></span>(counts_filtered, motif_ix[,<span class="kw">c</span>(<span class="dv">83</span>,<span class="dv">24</span>,<span class="dv">20</span>)])</code></pre></div>
<pre><code>## Warning: 'get_annotation_synergy' is deprecated.
## Use 'getAnnotationSynergy' instead.
## See help("Deprecated")</code></pre>
<pre><code>##                      MA0139.1_CTCF MA0107.1_RELA MA0140.2_GATA1::TAL1
## MA0139.1_CTCF            0.0000000     -2.001218            0.3338846
## MA0107.1_RELA           -2.0012182      0.000000            0.7990940
## MA0140.2_GATA1::TAL1     0.3338846      0.799094            0.0000000</code></pre>
<p>The result is a matrix with Z-scores for the variability synergy of each possible pairing. Note that this function is pretty slow and should only be performed with a limited selection of motifs and not the full collection!</p>
</div>
<div id="correlation" class="section level2">
<h2 class="hasAnchor">
<a href="#correlation" class="anchor"></a>Correlation</h2>
<p>An extreme synergy score suggests the possibility of cooperative or competitive binding between the factors that bind the motif. However, if the two factors simply tend to be co-expressed or oppositely expressed, then the synergy score could also be extreme. It can thus be helpful to look at the correlation between the deviations for the two motifs, in particular the correlation between the deviations for the peak sets that only have one or only the other motif. There is a function to compute this correlation:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/getAnnotationCorrelation.html">getAnnotationCorrelation</a></span>(counts_filtered, motif_ix[,<span class="kw">c</span>(<span class="dv">83</span>,<span class="dv">24</span>,<span class="dv">20</span>)])</code></pre></div>
<pre><code>## Warning: 'get_background_peaks' is deprecated.
## Use 'getBackgroundPeaks' instead.
## See help("Deprecated")</code></pre>
<pre><code>## Warning: 'compute_expectations' is deprecated.
## Use 'computeExpectations' instead.
## See help("Deprecated")</code></pre>
<pre><code>##                      MA0139.1_CTCF MA0107.1_RELA MA0140.2_GATA1::TAL1
## MA0139.1_CTCF            1.0000000   -0.41950962          -0.27827092
## MA0107.1_RELA           -0.4195096    1.00000000          -0.08943439
## MA0140.2_GATA1::TAL1    -0.2782709   -0.08943439           1.00000000</code></pre>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">Sys.Date</span>()</code></pre></div>
<pre><code>## [1] "2017-09-28"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.4.1 (2017-06-30)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04.2 LTS
## 
## Matrix products: default
## BLAS: /usr/lib/libblas/libblas.so.3.6.0
## LAPACK: /usr/lib/lapack/liblapack.so.3.6.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] TFBSTools_1.14.2                  ggmotif_0.0.0                    
##  [3] pheatmap_1.0.8                    BSgenome.Hsapiens.UCSC.hg19_1.4.0
##  [5] BSgenome_1.44.2                   rtracklayer_1.36.4               
##  [7] Biostrings_2.44.2                 XVector_0.16.0                   
##  [9] BiocParallel_1.10.1               ggplot2_2.2.1                    
## [11] Matrix_1.2-10                     SummarizedExperiment_1.6.4       
## [13] DelayedArray_0.2.7                matrixStats_0.52.2               
## [15] Biobase_2.36.2                    GenomicRanges_1.28.5             
## [17] GenomeInfoDb_1.12.2               IRanges_2.10.3                   
## [19] S4Vectors_0.14.4                  BiocGenerics_0.22.0              
## [21] motifmatchr_0.99.8                chromVAR_0.99.0                  
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                DirichletMultinomial_1.18.0
##  [3] bit64_0.9-7                 RColorBrewer_1.1-2         
##  [5] httr_1.3.1                  rprojroot_1.2              
##  [7] tools_3.4.1                 backports_1.1.0            
##  [9] R6_2.2.2                    DT_0.2                     
## [11] seqLogo_1.42.0              DBI_0.7                    
## [13] lazyeval_0.2.0              colorspace_1.3-2           
## [15] bit_1.1-12                  compiler_3.4.1             
## [17] plotly_4.7.1                labeling_0.3               
## [19] caTools_1.17.1              scales_0.5.0               
## [21] readr_1.1.1                 nabor_0.4.7                
## [23] stringr_1.2.0               digest_0.6.12              
## [25] Rsamtools_1.28.0            rmarkdown_1.6              
## [27] R.utils_2.5.0               pkgconfig_2.0.1            
## [29] htmltools_0.3.6             htmlwidgets_0.9            
## [31] rlang_0.1.2                 RSQLite_2.0                
## [33] VGAM_1.0-4                  shiny_1.0.3                
## [35] bindr_0.1                   jsonlite_1.5               
## [37] gtools_3.5.0                dplyr_0.7.2.9000           
## [39] R.oo_1.21.0                 RCurl_1.95-4.8             
## [41] magrittr_1.5                GO.db_3.4.1                
## [43] GenomeInfoDbData_0.99.0     Rcpp_0.12.12               
## [45] munsell_0.4.3               R.methodsS3_1.7.1          
## [47] stringi_1.1.5               yaml_2.1.14                
## [49] zlibbioc_1.22.0             Rtsne_0.13                 
## [51] plyr_1.8.4                  grid_3.4.1                 
## [53] blob_1.1.0                  miniUI_0.1.1               
## [55] CNEr_1.12.1                 lattice_0.20-35            
## [57] splines_3.4.1               annotate_1.54.0            
## [59] hms_0.3                     KEGGREST_1.16.1            
## [61] knitr_1.16                  JASPAR2016_1.4.0           
## [63] codetools_0.2-15            reshape2_1.4.2             
## [65] TFMPvalue_0.0.6             XML_3.98-1.9               
## [67] glue_1.1.1                  evaluate_0.10.1            
## [69] data.table_1.10.4           png_0.1-7                  
## [71] httpuv_1.3.5                tidyr_0.6.3                
## [73] purrr_0.2.3                 gtable_0.2.0               
## [75] poweRlaw_0.70.1             assertthat_0.2.0           
## [77] mime_0.5                    xtable_1.8-2               
## [79] viridisLite_0.2.0           tibble_1.3.4               
## [81] GenomicAlignments_1.12.2    AnnotationDbi_1.38.2       
## [83] memoise_1.1.0               bindrcpp_0.2</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#applications">Applications</a><ul class="nav nav-pills nav-stacked">
<li><a href="#variability">Variability</a></li>
      <li><a href="#clustering">Clustering</a></li>
      <li><a href="#cell-sample-similarity">Cell / sample similarity</a></li>
      <li><a href="#differential-accessibility-and-variability">Differential accessibility and variability</a></li>
      <li><a href="#motif-kmer-similarity">Motif / kmer similarity</a></li>
      <li><a href="#kmers-and-sequence-specificity-of-variation">Kmers and sequence specificity of variation</a></li>
      <li><a href="#de-novo-kmer-assembly">De novo kmer assembly</a></li>
      <li><a href="#synergy-between-motifs">Synergy between motifs</a></li>
      <li><a href="#function-for-computing-synergy">Function for computing synergy</a></li>
      <li><a href="#correlation">Correlation</a></li>
      <li><a href="#session-info">Session Info</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Alicia Schep.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
