<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title> &bull; chromVAR</title><!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous"><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet"><script src="../../jquery.sticky-kit.min.js"></script><script src="../../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../../index.html">chromVAR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../../articles/Introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../../articles/Articles/Counts.html">Counts</a>
    </li>
    <li>
      <a href="../../articles/Articles/Annotations.html">Annotations</a>
    </li>
    <li>
      <a href="../../articles/Articles/Deviations.html">Deviations</a>
    </li>
    <li>
      <a href="../../articles/Articles/Applications.html">Applications</a>
    </li>
  </ul></li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/GreenleafLab/chromVAR">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1></h1>
            
          </div>

    
    
<div class="contents">
<div id="counts" class="section level1">
<h1 class="hasAnchor"><html><body><a href="#counts" class="anchor"> </a></body></html>Counts</h1>
<p>This vignette covers the correct format for storing fragment counts for use in chromVAR.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(chromVAR)
<span class="kw">library</span>(SummarizedExperiment)
<span class="kw">library</span>(Matrix)</code></pre></div>
<div id="chromvars-method-for-getting-fragment-counts" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#chromvars-method-for-getting-fragment-counts" class="anchor"> </a></body></html>chromVAR&rsquo;s method for getting fragment counts</h2>
<p>chromVAR includes a function <code>get_counts</code> for reading in fragment counts from bam or bed files.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Not evaluated -- Fake files!</span>
peakfile &lt;-<span class="st"> "mypeaks.bed"</span>
peaks &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/get_peaks.html">get_peaks</a></span>(peakfile)

bamfiles &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"mybam1.bam"</span>,<span class="st">"mybam2.bam"</span>)
fragment_counts &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/get_counts.html">get_counts</a></span>(bamfiles, peaks, <span class="dt">paired =</span>  <span class="ot">TRUE</span>, <span class="dt">by_rg =</span> <span class="ot">TRUE</span>, <span class="dt">format =</span> <span class="st">"bam"</span>, 
                              <span class="dt">colData =</span> <span class="kw">DataFrame</span>(<span class="dt">celltype =</span> <span class="kw">c</span>(<span class="st">"GM"</span>,<span class="st">"K562"</span>)))</code></pre></div>
<div id="rg-tags" class="section level3">
<h3 class="hasAnchor"><html><body><a href="#rg-tags" class="anchor"> </a></body></html>RG tags</h3>
<p>The <code>by_rg</code> argument indicates that RG tags are used to distinguish reads from different cells or samples. If RG tags are not used and each bam file represents an individual sample, use <code>by_rg = FALSE</code>. If the format of the files is &ldquo;bed&rdquo;, use <code>format = bed</code> instead of <code>bam</code>.</p>
</div>
<div id="paired-vs-single-end-data" class="section level3">
<h3 class="hasAnchor"><html><body><a href="#paired-vs-single-end-data" class="anchor"> </a></body></html>Paired vs single end data</h3>
<p>The <code>paired</code> argument should be set based on whether the data is paired end. With paired end data, fragments are counted once if either or both 5&rsquo; ends of the fragment map to a peak. For single end data, a fragment is counted if the 5&rsquo; end maps to the peak.</p>
</div>
<div id="sample-annotation" class="section level3">
<h3 class="hasAnchor"><html><body><a href="#sample-annotation" class="anchor"> </a></body></html>Sample annotation</h3>
<p>The colData argument takes in a DataFrame with sample annotations. If <code>by_rg = TRUE</code> and the number of rows in the DataFrame is equal to the number of alignment files rather than the number of cells/samples indicated by the RG tags, then the annotations are propagated to all cells/samples in each alignment file. In the example above, all cells in &ldquo;mybam1.bam&rdquo; would get assigned celltype as &ldquo;GM&rdquo; and all cells in &ldquo;mybam2.bam&rdquo; would get assigned as cell type &ldquo;K562&rdquo;.</p>
</div>
<div id="sequencing-depth" class="section level3">
<h3 class="hasAnchor"><html><body><a href="#sequencing-depth" class="anchor"> </a></body></html>Sequencing depth</h3>
<p>When chromVAR reads in alignments to determine fragment counts per peak, it also reads in the overall sequencing depth for each sample. This information is stored in the <code>colData</code> of the resulting SummarizedExperiment as the column &ldquo;depth&rdquo;.</p>
</div>
</div>
<div id="using-your-own-matrix-of-fragment-counts" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#using-your-own-matrix-of-fragment-counts" class="anchor"> </a></body></html>Using your own matrix of fragment counts</h2>
<p>If you have already computed a matrix of fragment counts per peaks and don&rsquo;t want to use chromVAR&rsquo;s <code>get_counts</code> function, then it is recommended to create your own RangedSummarizedExperiment object that matches the output one would get from <code>get_counts</code>. See the documentation for SummarizedExperiment for more information on the (Ranged)SummarizedExperiment constructor function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Not evaluated -- example (not real) objects</span>
fragment_counts &lt;-<span class="st"> </span><span class="kw">SummarizedExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> my_counts_matrix),
                                        <span class="dt">colData =</span> sample_annotation,
                                        <span class="dt">rowRanges =</span> peaks)</code></pre></div>
<p>In order to be able to use the <code>filter_samples</code> function, a &ldquo;depth&rdquo; column with the total sequencing depth must be included in the <code>colData</code> in the SummarizedExperiment object.</p>
<p>It is possible to instead use just a simple matrix or Matrix as input to the main <code>compute_deviations</code> function. However, you will have to first compute background peaks using <code>get_background_peaks</code> providing your own computed bias vector.</p>
</div>
<div id="adding-gc-content" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#adding-gc-content" class="anchor"> </a></body></html>Adding GC content</h2>
<p>If your counts are stored as a RangedSummarizedExperiment, then gc bias can be added to rowRanges data. The GC content will be used for determining background peaks. The function <code>add_gc_bias</code> returns an updated RangedSummarizedExperiment with a new rowData column named &ldquo;bias&rdquo;.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(example_counts)
example_counts &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/add_gc_bias.html">add_gc_bias</a></span>(example_counts)
<span class="kw">head</span>(<span class="kw">rowData</span>(example_counts))</code></pre></div>
<pre><code>## DataFrame with 6 rows and 4 columns
##       score      qval        name      bias
##   &lt;integer&gt; &lt;numeric&gt; &lt;character&gt; &lt;numeric&gt;
## 1       259  25.99629   GM_peak_6     0.652
## 2        82   8.21494   H1_peak_7     0.680
## 3       156  15.65456  GM_peak_17     0.788
## 4        82   8.21494  H1_peak_13     0.674
## 5       140  14.03065  GM_peak_27     0.600
## 6       189  18.99529 GM_peak_29a     0.742</code></pre>
<p>Note that the function <code>add_gc_bias</code> also takes in an argument for a BSgenome object. The default is BSgenome.Hsapiens.UCSC.hg19, so if using a different genome build be sure to provide the correct genome. For example, if using sacCer3 you could do:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Not evaluated -- wrong genome</span>
<span class="kw">library</span>(BSgenome.Scerevisiae.UCSC.sacCer3)
example_counts &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/add_gc_bias.html">add_gc_bias</a></span>(example_counts, <span class="dt">genome =</span> BSgenome.Scerevisiae.UCSC.sacCer3)</code></pre></div>
<p>Check out <code>available.genomes</code> from the BSgenome package for what genomes are available. For making your own BSgenome object, check out <code>BSgenomeForge</code>.</p>
</div>
<div id="filtering-cells" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#filtering-cells" class="anchor"> </a></body></html>Filtering cells</h2>
<p>If working with single cell data, it is advisable to filter out samples with insufficient reads or a low proportion of reads in peaks as these may represent empty wells or dead cells. Two parameters are used for filtering &ndash; min_in_peaks and min_depth. If not provided (as above), these cutoffs are estimated based on the medians from the data. min_in_peaks is set to 0.5 times the median proportion of fragments in peaks. min_depth is set to the maximum of 500 or 10% of the median library size.</p>
<p>Unless <code>plot = FALSE</code> given as argument to function <code>filter_samples</code>, a plot will be generated.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/filter_samples.html">filter_samples</a></span>(example_counts, <span class="dt">min_depth =</span> <span class="dv">1500</span>, <span class="dt">min_in_peaks =</span> <span class="fl">0.15</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>If shiny argument is set to TRUE (the default), a shiny gadget will pop up which allows you to play with the filtering parameters and see which cells pass filters or not.</p>
<p>To get just the plot of what is filtered, use <code>filter_samples_plot</code>. By default, the plot is interactive&ndash; to set it as not interactive use <code>use_plotly = FALSE</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filtering_plot &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/filter_samples_plot.html">filter_samples_plot</a></span>(example_counts, <span class="dt">min_depth =</span> <span class="dv">1500</span>, <span class="dt">min_in_peaks =</span> <span class="fl">0.15</span>, <span class="dt">use_plotly =</span> <span class="ot">FALSE</span>)
filtering_plot</code></pre></div>
<p><img src="Counts_files/figure-html/filter_plot-1.png" width="576"></p>
<p>To instead return the indexes of the samples to keep instead of a new SummarizedExperiment object, use ix_return = TRUE.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ix &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/filter_samples.html">filter_samples</a></span>(example_counts, <span class="dt">ix_return =</span> <span class="ot">TRUE</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="filtering-peaks" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#filtering-peaks" class="anchor"> </a></body></html>Filtering peaks</h2>
<p>For both bulk and single cell data, peaks should be filtered based on having at least a certain number of fragments. At minimum, each peak should have at least one fragment across all the samples (it might be possible to have peaks with zero reads due to using a peak set defined by other data). Otherwise, downstream functions won&rsquo;t work. The function <code>filter_peaks</code> will also reduce the peak set to non-overlapping peaks (keeping the peak with higher counts for peaks that overlap) if non_overlapping argument is set to TRUE (which is default).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/filter_peaks.html">filter_peaks</a></span>(counts_filtered, <span class="dt">non_overlapping =</span> <span class="ot">TRUE</span>)</code></pre></div>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#session-info" class="anchor"> </a></body></html>Session Info</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">Sys.Date</span>()</code></pre></div>
<pre><code>## [1] "2017-02-07"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.3.1 (2016-06-21)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.1 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] chromVAR_0.4.1             TFBSTools_1.12.0          
##  [3] ggmotif_0.0.0              pheatmap_1.0.8            
##  [5] ggplot2_2.2.0              BiocParallel_1.8.0        
##  [7] SummarizedExperiment_1.4.0 Biobase_2.34.0            
##  [9] GenomicRanges_1.26.1       GenomeInfoDb_1.10.0       
## [11] IRanges_2.8.0              S4Vectors_0.12.0          
## [13] BiocGenerics_0.20.0        Matrix_1.2-7.1            
## [15] motifmatchr_0.1.0          shiny_0.14.2              
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                      DirichletMultinomial_1.16.0      
##  [3] devtools_1.12.0.9000              RColorBrewer_1.1-2               
##  [5] httr_1.2.1                        rprojroot_1.1                    
##  [7] highlight_0.4.7                   tools_3.3.1                      
##  [9] backports_1.0.4                   R6_2.2.0                         
## [11] DT_0.2                            seqLogo_1.40.0                   
## [13] DBI_0.5-1                         lazyeval_0.2.0                   
## [15] colorspace_1.2-7                  withr_1.0.2                      
## [17] xml2_1.0.0.9001                   desc_1.0.1                       
## [19] plotly_4.5.6                      labeling_0.3                     
## [21] rtracklayer_1.34.0                caTools_1.17.1                   
## [23] scales_0.4.1                      readr_1.0.0                      
## [25] callr_1.0.0.9000                  pkgdown_0.1.0.9000               
## [27] nabor_0.4.6                       stringr_1.1.0                    
## [29] digest_0.6.10                     Rsamtools_1.26.1                 
## [31] rmarkdown_1.3                     R.utils_2.5.0                    
## [33] BSgenome.Hsapiens.UCSC.hg19_1.4.0 XVector_0.14.0                   
## [35] base64enc_0.1-3                   htmltools_0.3.5                  
## [37] BSgenome_1.42.0                   htmlwidgets_0.8                  
## [39] RSQLite_1.1-1                     VGAM_1.0-2                       
## [41] jsonlite_1.1                      gtools_3.5.0                     
## [43] dplyr_0.5.0                       R.oo_1.21.0                      
## [45] RCurl_1.95-4.8                    magrittr_1.5                     
## [47] GO.db_3.4.0                       Rcpp_0.12.8                      
## [49] munsell_0.4.3                     R.methodsS3_1.7.1                
## [51] stringi_1.1.2                     whisker_0.3-2                    
## [53] yaml_2.1.14                       zlibbioc_1.20.0                  
## [55] Rtsne_0.11                        pkgbuild_0.0.0.9000              
## [57] plyr_1.8.4                        grid_3.3.1                       
## [59] crayon_1.3.2                      miniUI_0.1.1                     
## [61] CNEr_1.10.0                       lattice_0.20-34                  
## [63] Biostrings_2.42.0                 splines_3.3.1                    
## [65] annotate_1.52.0                   knitr_1.15.1                     
## [67] JASPAR2016_1.2.0                  codetools_0.2-15                 
## [69] reshape2_1.4.2                    pkgload_0.0.0.9000               
## [71] TFMPvalue_0.0.6                   XML_3.98-1.5                     
## [73] evaluate_0.10                     RcppArmadillo_0.7.600.1.0        
## [75] httpuv_1.3.3                      testthat_1.0.2                   
## [77] gtable_0.2.0                      poweRlaw_0.60.3                  
## [79] purrr_0.2.2                       tidyr_0.6.0                      
## [81] assertthat_0.1                    mime_0.5                         
## [83] xtable_1.8-2                      roxygen2_5.0.1                   
## [85] viridisLite_0.1.3                 tibble_1.2                       
## [87] GenomicAlignments_1.10.0          AnnotationDbi_1.36.0             
## [89] memoise_1.0.0</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked"><li><a href="#counts">Counts</a><ul class="nav nav-pills nav-stacked"><li><a href="#chromvars-method-for-getting-fragment-counts">chromVAR&rsquo;s method for getting fragment counts</a></li>
      <li><a href="#using-your-own-matrix-of-fragment-counts">Using your own matrix of fragment counts</a></li>
      <li><a href="#adding-gc-content">Adding GC content</a></li>
      <li><a href="#filtering-cells">Filtering cells</a></li>
      <li><a href="#filtering-peaks">Filtering peaks</a></li>
      <li><a href="#session-info">Session Info</a></li>
      </ul></li>
      </ul></div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Alicia Schep.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer></div>

  </body></html>
