<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • chromVAR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../jquery.sticky-kit.min.js"></script><script src="../../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../../index.html">chromVAR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/Introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../../articles/Articles/Counts.html">Counts</a>
    </li>
    <li>
      <a href="../../articles/Articles/Annotations.html">Annotations</a>
    </li>
    <li>
      <a href="../../articles/Articles/Deviations.html">Deviations</a>
    </li>
    <li>
      <a href="../../articles/Articles/Applications.html">Applications</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/GreenleafLab/chromVAR">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1></h1>
            
          </div>

    
    
<div class="contents">
<div id="deviations" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#deviations" class="anchor"> </a></body></html>Deviations</h1>
<p>This vignette covers the main function of chromVAR, compute_deviations.</p>
<div id="inputs" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#inputs" class="anchor"> </a></body></html>Inputs</h2>
<p>The <code>compute_deviations</code> has two required inputs, <code>object</code> and <code>annotations</code>. The first argument should be a RangedSummarizedExperiment with a <code>counts</code> assay storing the fragment counts per peak (rows) per cell or sample (columns). The second argument (annnotations) argument should be a RangedSummarizedExperiment with the <code>matches</code> assay storing whether each peak (rows) contains each annotation (columns). For more information on these two primary inputs, see the “Counts” and “Annotations” vignettes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Getting inputs ready -- See Inputs vignette for more info</span>
<span class="kw">library</span>(chromVAR)
<span class="kw">library</span>(motifmatchr)
<span class="kw">library</span>(SummarizedExperiment)
<span class="kw">library</span>(Matrix)
<span class="kw">data</span>(example_counts, <span class="dt">package =</span> <span class="st">"chromVAR"</span>)
example_counts &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/add_gc_bias.html">add_gc_bias</a></span>(example_counts)
counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/filter_samples.html">filter_samples</a></span>(example_counts, <span class="dt">min_depth =</span> <span class="dv">1500</span>, <span class="dt">min_in_peaks =</span> <span class="fl">0.15</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)
counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/filter_peaks.html">filter_peaks</a></span>(counts_filtered)
motifs &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/get_jaspar_motifs.html">get_jaspar_motifs</a></span>()
motif_ix &lt;-<span class="st"> </span><span class="kw">match_motifs</span>(motifs, counts_filtered)

<span class="co"># computing deviations</span>
dev &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/compute_deviations.html">compute_deviations</a></span>(<span class="dt">object =</span> counts_filtered, 
                                 <span class="dt">annotations =</span> motif_ix)</code></pre></div>
<p>Optionally, the annotations can be omitted and then deviations will be computed per peak. That approach should only be use with bulk, deeply sequenced data.</p>
</div>
<div id="outputs" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#outputs" class="anchor"> </a></body></html>Outputs</h2>
<p>The output from the compute_deviations function is a chromVARDeviations object. This object inherits from the RangedSummarizedExperiment object, so anything you can do with that object can be done with this one.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(dev)</code></pre></div>
<pre><code>## class: chromVARDeviations 
## dim: 386 36 
## metadata(0):
## assays(2): deviations z
## rownames(386): MA0025.1_NFIL3 MA0030.1_FOXF2 ... MA0909.1_HOXD13
##   MA0914.1_ISL2
## rowData names(2): pwm name
## colnames(36): singles-GM12878-140905-1 singles-GM12878-140905-2
##   ... singles-H1ESC-140820-24 singles-H1ESC-140820-25
## colData names(2): Cell_Type depth</code></pre>
<p>The object stores two main ‘assays’, which can be accessed using the deviations and deviation_scores functions. The deviations are the bias corrected deviations in accessibility. For each motif or annotation (rows), there is a value for each cell or sample (columns) representing how different the accessibility for peaks with that motif or annotation is from the expectation based on all cells being equal, corrected for biases.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/deviations.html">deviations</a></span>(dev)[<span class="dv">1</span>:<span class="dv">3</span>,<span class="dv">1</span>:<span class="dv">2</span>]</code></pre></div>
<pre><code>##                singles-GM12878-140905-1 singles-GM12878-140905-2
## MA0025.1_NFIL3              -0.06253114               0.11849419
## MA0030.1_FOXF2              -0.05561553              -0.07242295
## MA0031.1_FOXD1              -0.08999824              -0.18741128</code></pre>
<p>The deviation_scores are the Z-scores for each bias corrected deviations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../../reference/deviation_scores.html">deviation_scores</a></span>(dev)[<span class="dv">1</span>:<span class="dv">3</span>,<span class="dv">1</span>:<span class="dv">2</span>]</code></pre></div>
<pre><code>##                singles-GM12878-140905-1 singles-GM12878-140905-2
## MA0025.1_NFIL3               -0.4112919                0.8738719
## MA0030.1_FOXF2               -0.4434714               -0.7905780
## MA0031.1_FOXD1               -0.7734451               -2.6886946</code></pre>
<p>The absolute value of the deviation scores will be correlated with the read depth as with more reads, there can be more confidence that the difference in accessibility from the expecation is greater than would occur by chance.</p>
</div>
<div id="options" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#options" class="anchor"> </a></body></html>Options</h2>
<p>The compute_deviations function also has two optional arguments, <code>background_peaks</code> and <code>expectation</code>.</p>
<div id="background-peaks" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#background-peaks" class="anchor"> </a></body></html>Background peaks</h3>
<p>The function compute_deviations will use a set of background peaks for bias correcting the deviations. This computation is done internally by default and not returned – to have greater control over this step, a user can run the get_background_peaks function themselves and pass the result to compute_deviations under the <code>background_peaks</code> parameter. If the fragment counts objects is a Matrix/matrix or a SummarizedExperiment without a “bias” column in the rowData, then it is mandatory to supply a background_peaks argument.</p>
<p>Background peaks are peaks that are similar to a peak in GC content and average accessibility.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bg &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/get_background_peaks.html">get_background_peaks</a></span>(<span class="dt">object =</span> counts_filtered)</code></pre></div>
<p>The result from get_background_peaks is a matrix of indices, where each column represents the index of the peak that is a background peak. The number of columns represents the number of background iterations that will be used for computing the bias corrected deviations.</p>
<p>To use the background peaks computed, simply add those to the call to compute_deviations:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dev &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/compute_deviations.html">compute_deviations</a></span>(<span class="dt">object =</span> counts_filtered, 
                                 <span class="dt">annotations =</span> motif_ix, 
                                 <span class="dt">background_peaks =</span> bg)</code></pre></div>
<p>If you are using a counts matrix or a SummarizedExperiment without a “bias” column in the rowRanges or rowData, then a bias argument should be given as well.</p>
</div>
<div id="expectation" class="section level3">
<h3 class="hasAnchor">
<html><body><a href="#expectation" class="anchor"> </a></body></html>Expectation</h3>
<p>By default, chromVAR measures the difference in chromatin accessibility for an annotation relative to the expectation if all cells or samples have the same chromatin accessibility profile but differ simply in read depth. Optionally, this expectation can be adjusted. The function ‘compute_expecations’ can calculate the expectation in alternate ways. The function can also be applied to only a subset of the cells or samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">expected &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/compute_expectations.html">compute_expectations</a></span>(counts_filtered) </code></pre></div>
<p>By default, this function will compute the expected fraction of reads per peak as the the total fragments per peak across all samples divided by total reads in peaks in all samples. Optionally, <code>norm</code> can be set to <code>TRUE</code> and then the expectation will be the average fraction of reads in a peak across the cells. This is <strong>not recommended</strong> for single cell applications as cells with very few reads will have a large impact.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Not recommended for single cell data or other very sparse data!</span>
expected &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/compute_expectations.html">compute_expectations</a></span>(counts_filtered, <span class="dt">norm =</span> <span class="ot">TRUE</span>) </code></pre></div>
<p>Another option is to give a vector of groups, in which case the expectation will be the average fraction of reads per peak within each group. If a group vector is provided and <code>norm</code> is set to <code>TRUE</code> then within each group the fraction of reads per peak is the average fraction of reads per peak in each sample. Otherwise, the within group fraction of reads per peak is based on the reads per peak within the sample divided by the total reads within each sample.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">expected &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/compute_expectations.html">compute_expectations</a></span>(counts_filtered, 
                                 <span class="dt">group =</span> <span class="kw">colData</span>(counts_filtered)$Cell_Type) </code></pre></div>
<p>To use your computed expectation vector, simply pass it to the <code>expectation</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dev &lt;-<span class="st"> </span><span class="kw"><a href="../../reference/compute_deviations.html">compute_deviations</a></span>(<span class="dt">object =</span> counts_filtered, 
                          <span class="dt">annotations =</span> motif_ix, 
                          <span class="dt">expectation =</span> expected)</code></pre></div>
</div>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#session-info" class="anchor"> </a></body></html>Session Info</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">Sys.Date</span>()</code></pre></div>
<pre><code>## [1] "2017-02-21"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.3.1 (2016-06-21)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.1 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] ggmotif_0.0.0              chromVAR_0.4.3            
##  [3] pheatmap_1.0.8             ggplot2_2.2.1             
##  [5] BiocParallel_1.8.1         Matrix_1.2-7.1            
##  [7] TFBSTools_1.12.1           shiny_1.0.0               
##  [9] SummarizedExperiment_1.4.0 Biobase_2.34.0            
## [11] GenomicRanges_1.26.2       GenomeInfoDb_1.10.3       
## [13] IRanges_2.8.1              S4Vectors_0.12.1          
## [15] BiocGenerics_0.20.0        motifmatchr_0.1.0         
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                      DirichletMultinomial_1.16.0      
##  [3] devtools_1.12.0.9000              RColorBrewer_1.1-2               
##  [5] httr_1.2.1                        rprojroot_1.2                    
##  [7] highlight_0.4.7                   backports_1.0.5                  
##  [9] tools_3.3.1                       R6_2.2.0                         
## [11] DT_0.2                            seqLogo_1.40.0                   
## [13] DBI_0.5-1                         lazyeval_0.2.0                   
## [15] colorspace_1.3-2                  withr_1.0.2                      
## [17] curl_2.3                          git2r_0.18.0                     
## [19] xml2_1.1.1                        desc_1.1.0                       
## [21] plotly_4.5.6                      labeling_0.3                     
## [23] rtracklayer_1.34.1                caTools_1.17.1                   
## [25] scales_0.4.1                      readr_1.0.0                      
## [27] callr_1.0.0.9000                  pkgdown_0.1.0.9000               
## [29] commonmark_1.1                    nabor_0.4.6                      
## [31] stringr_1.1.0                     digest_0.6.12.1                  
## [33] Rsamtools_1.26.1                  rmarkdown_1.3                    
## [35] R.utils_2.5.0                     BSgenome.Hsapiens.UCSC.hg19_1.4.0
## [37] XVector_0.14.0                    base64enc_0.1-3                  
## [39] htmltools_0.3.5                   BSgenome_1.42.0                  
## [41] htmlwidgets_0.8                   RSQLite_1.1-2                    
## [43] VGAM_1.0-3                        BiocInstaller_1.24.0             
## [45] jsonlite_1.2                      gtools_3.5.0                     
## [47] dplyr_0.5.0                       R.oo_1.21.0                      
## [49] RCurl_1.95-4.8                    magrittr_1.5                     
## [51] GO.db_3.4.0                       Rcpp_0.12.9                      
## [53] munsell_0.4.3                     R.methodsS3_1.7.1                
## [55] whisker_0.3-2                     yaml_2.1.14                      
## [57] stringi_1.1.2                     zlibbioc_1.20.0                  
## [59] Rtsne_0.11                        pkgbuild_0.0.0.9000              
## [61] plyr_1.8.4                        grid_3.3.1                       
## [63] crayon_1.3.2                      miniUI_0.1.1                     
## [65] CNEr_1.10.1                       lattice_0.20-34                  
## [67] Biostrings_2.42.1                 splines_3.3.1                    
## [69] annotate_1.52.1                   KEGGREST_1.14.0                  
## [71] knitr_1.15.1                      JASPAR2016_1.2.0                 
## [73] codetools_0.2-15                  reshape2_1.4.2                   
## [75] pkgload_0.0.0.9000                TFMPvalue_0.0.6                  
## [77] XML_3.98-1.5                      evaluate_0.10                    
## [79] RcppArmadillo_0.7.700.0.0         png_0.1-7                        
## [81] httpuv_1.3.3                      testthat_1.0.2                   
## [83] gtable_0.2.0                      poweRlaw_0.70.0                  
## [85] purrr_0.2.2                       tidyr_0.6.1                      
## [87] assertthat_0.1                    mime_0.5                         
## [89] xtable_1.8-2                      roxygen2_6.0.1                   
## [91] viridisLite_0.1.3                 tibble_1.2                       
## [93] GenomicAlignments_1.10.0          AnnotationDbi_1.36.2             
## [95] memoise_1.0.0</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#deviations">Deviations</a><ul class="nav nav-pills nav-stacked">
<li><a href="#inputs">Inputs</a></li>
      <li><a href="#outputs">Outputs</a></li>
      <li><a href="#options">Options</a></li>
      <li><a href="#session-info">Session Info</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Alicia Schep.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
