<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Chromatin Variation Across Regions â€¢ chromVAR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">chromVAR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Introduction.html">Introduction</a>
    </li>
    <li>
      <a href="articles/Articles/Counts.html">Counts</a>
    </li>
    <li>
      <a href="articles/Articles/Annotations.html">Annotations</a>
    </li>
    <li>
      <a href="articles/Articles/Deviations.html">Deviations</a>
    </li>
    <li>
      <a href="articles/Articles/Applications.html">Applications</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/GreenleafLab/chromVAR">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    
<hr>
<div id="chromvar" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#chromvar" class="anchor"> </a>chromVAR</h1></div>
<p>chromVAR is an R package for the analysis of sparse chromatin accessibility data from single cell or bulk ATAC or DNAse-seq data.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"> </a>Installation</h2>
<p>Installation is easiest using the devtools package. The function <code>install_github</code> will install the package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">"GreenleafLab/chromVAR"</span>)</code></pre></div>
<p>Two additional packages that are recommended and used in the vignettes:</p>
<ul>
<li>motifmatchr - available on <a href="https://github.com/GreenleafLab/motifmatchr">GitHub</a>
</li>
<li>JASPAR2016 - available from Bioconductor</li>
</ul>
<p>A number of needed packages are installed in this process. One of the dependencies has a system requirement for the gsl library, so if this is not installed already it may need to be installed separately.</p>
</div>
<div id="quickstart" class="section level2">
<h2 class="hasAnchor">
<a href="#quickstart" class="anchor"> </a>Quickstart</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(chromVAR)
<span class="kw">library</span>(motifmatchr)

### Example of how to read in counts -------------------------------------------

<span class="co"># Caution: FAKE FILENAMES -- Replace with real as appropriate! If you want to </span>
<span class="co"># run on example data in package, start at next section with example_counts data</span>

peakfile &lt;-<span class="st"> "mypeaks.bed"</span>
peaks &lt;-<span class="st"> </span><span class="kw"><a href="reference/get_peaks.html">get_peaks</a></span>(peakfile)

bamfiles &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"mybam1.bam"</span>,<span class="st">"mybam2.bam"</span>)
fragment_counts &lt;-<span class="st"> </span><span class="kw"><a href="reference/get_counts.html">get_counts</a></span>(bamfiles, peaks, 
                              <span class="dt">paired =</span>  <span class="ot">TRUE</span>, 
                              <span class="dt">by_rg =</span> <span class="ot">TRUE</span>, 
                              <span class="dt">format =</span> <span class="st">"bam"</span>, 
                              <span class="dt">colData =</span> <span class="kw">DataFrame</span>(<span class="dt">celltype =</span> <span class="kw">c</span>(<span class="st">"GM"</span>,<span class="st">"K562"</span>)))

### ----------------------------------------------------------------------------

### Using example counts from package ------------------------------------------

<span class="kw">data</span>(example_counts, <span class="dt">package =</span> <span class="st">"chromVAR"</span>)
example_counts &lt;-<span class="st"> </span><span class="kw"><a href="reference/add_gc_bias.html">add_gc_bias</a></span>(example_counts)
counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="reference/filter_samples.html">filter_samples</a></span>(example_counts, <span class="dt">min_depth =</span> <span class="dv">1500</span>,
                                  <span class="dt">min_in_peaks =</span> <span class="fl">0.15</span>)
counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="reference/filter_peaks.html">filter_peaks</a></span>(counts_filtered)
motifs &lt;-<span class="st"> </span><span class="kw"><a href="reference/get_jaspar_motifs.html">get_jaspar_motifs</a></span>()
motif_ix &lt;-<span class="st"> </span><span class="kw">match_motifs</span>(motifs, counts_filtered)

<span class="co"># computing deviations</span>
dev &lt;-<span class="st"> </span><span class="kw"><a href="reference/compute_deviations.html">compute_deviations</a></span>(<span class="dt">object =</span> counts_filtered, 
                                 <span class="dt">annotations =</span> motif_ix)</code></pre></div>
<p>See Vignettes for more information!</p>
</div>
</div>

  </div>

  <div class="col-md-3" id="sidebar">
    <h2>License</h2>
<p><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE">LICENSE</a></p>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Alicia Schep <br><small class="roles"> Author, maintainer </small> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Alicia Schep.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
