<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Chromatin Variation Across Regions • chromVAR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">chromVAR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Introduction.html">Introduction</a>
    </li>
    <li>
      <a href="articles/Articles/Counts.html">Counts</a>
    </li>
    <li>
      <a href="articles/Articles/Annotations.html">Annotations</a>
    </li>
    <li>
      <a href="articles/Articles/Deviations.html">Deviations</a>
    </li>
    <li>
      <a href="articles/Articles/Applications.html">Applications</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/GreenleafLab/chromVAR">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    
<hr>
<div id="chromvar" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#chromvar" class="anchor"></a>chromVAR</h1></div>
<p>chromVAR is an R package for the analysis of sparse chromatin accessibility data from single cell or bulk ATAC or DNAse-seq data.</p>
<div id="note-on-recent-function-name-changes" class="section level2">
<h2 class="hasAnchor">
<a href="#note-on-recent-function-name-changes" class="anchor"></a>Note on recent function name changes</h2>
<p>The chromVAR (and related motifmatchr) function names recently changed to switch over to camelCase from snake_case. All exported functions now use camelCase, e.g. <code>compute_deviations</code> is now <code>computeDeviations</code>. If following the current documentation but using an earlier version of the package, either update the package or be aware of the discrepancy. This change was made to comply with Bioconductor naming preferences, as chromVAR will soon be submitted to Bioconductor.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>The recommended installation method for <code>chromVAR</code> is using the <code>biocLite</code> command from BiocInstaller. You will first have to have installed the <a href="https://bioconductor.org/packages/release/bioc/html/BiocInstaller.html">BiocInstaller package</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">BiocInstaller::<span class="kw">biocLite</span>(<span class="st">"GreenleafLab/chromVAR"</span>)</code></pre></div>
<p>A number of needed packages are installed in this process. One of the dependencies has a system requirement for the gsl library, so if this is not installed already it may need to be installed separately.</p>
<p>For Windows users, some have reported that the S4Vector dependency does not currently function on windows R 3.3.3, but that installation was successful on R 3.3.2</p>
<p>Two additional packages that are recommended and used in the vignettes:</p>
<ul>
<li>motifmatchr - available on <a href="https://github.com/GreenleafLab/motifmatchr">GitHub</a> or <a href="https://bioconductor.org/packages/devel/bioc/html/motifmatchr.html">development version of Bioconductor</a>
</li>
<li>JASPAR2016 - available from Bioconductor</li>
</ul>
<p>Additionally, the package chromVARmotifs can be useful for loading additional motif collections:</p>
<ul>
<li>chromVARmotifs - available on <a href="https://github.com/GreenleafLab/chromVARmotifs">GitHub</a>
</li>
</ul>
</div>
<div id="parallelization" class="section level2">
<h2 class="hasAnchor">
<a href="#parallelization" class="anchor"></a>Parallelization</h2>
<p>Before running chromVAR functions, it is advisable to use the <code>register</code> function from BiocParallel to specify your preferred method of parallelization. For unix systems,</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(BiocParallel)
<span class="kw">register</span>(<span class="kw">MulticoreParam</span>(<span class="dv">8</span>)) <span class="co"># Use 8 cores</span></code></pre></div>
<p>For Windows, <code>MulticoreParam</code> will not work, but you can use SnowParam:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">register</span>(<span class="kw">SnowParam</span>(<span class="kw">SnowParam</span>(<span class="dt">workers=</span><span class="dv">1</span>, <span class="dt">type =</span> <span class="st">"SOCK"</span>)))</code></pre></div>
<p>Even if you don’t want to use more than one core, it is recommended to explicitly register that choice using SerialParam:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">register</span>(<span class="kw">SerialParam</span>())</code></pre></div>
<p>Please see the documentation for <a href="https://bioconductor.org/packages/release/bioc/html/BiocParallel.html">BiocParallel</a> for more information about the <code>register</code> function and the various options for multi-processing.</p>
</div>
<div id="quickstart" class="section level2">
<h2 class="hasAnchor">
<a href="#quickstart" class="anchor"></a>Quickstart</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(chromVAR)
<span class="kw">library</span>(motifmatchr)
<span class="kw">library</span>(BSgenome.Hsapiens.UCSC.hg19)

### Example of how to read in counts -------------------------------------------

<span class="co"># Caution: FAKE FILENAMES -- Replace with real as appropriate! If you want to </span>
<span class="co"># run on example data in package, start at next section with example_counts data</span>

peakfile &lt;-<span class="st"> "mypeaks.bed"</span>
peaks &lt;-<span class="st"> </span><span class="kw"><a href="reference/getPeaks.html">getPeaks</a></span>(peakfile)

bamfiles &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"mybam1.bam"</span>,<span class="st">"mybam2.bam"</span>)
fragment_counts &lt;-<span class="st"> </span><span class="kw"><a href="reference/getCounts.html">getCounts</a></span>(bamfiles, peaks, 
                              <span class="dt">paired =</span>  <span class="ot">TRUE</span>, 
                              <span class="dt">by_rg =</span> <span class="ot">TRUE</span>, 
                              <span class="dt">format =</span> <span class="st">"bam"</span>, 
                              <span class="dt">colData =</span> <span class="kw">DataFrame</span>(<span class="dt">celltype =</span> <span class="kw">c</span>(<span class="st">"GM"</span>,<span class="st">"K562"</span>)))

### ----------------------------------------------------------------------------

### Using example counts from package ------------------------------------------

<span class="kw">data</span>(example_counts, <span class="dt">package =</span> <span class="st">"chromVAR"</span>)
example_counts &lt;-<span class="st"> </span><span class="kw"><a href="reference/addGCBias.html">addGCBias</a></span>(example_counts, 
                              <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19)
counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="reference/filterSamples.html">filterSamples</a></span>(example_counts, <span class="dt">min_depth =</span> <span class="dv">1500</span>,
                                  <span class="dt">min_in_peaks =</span> <span class="fl">0.15</span>)
counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="reference/filterPeaks.html">filterPeaks</a></span>(counts_filtered)
motifs &lt;-<span class="st"> </span><span class="kw"><a href="reference/getJasparMotifs.html">getJasparMotifs</a></span>()
motif_ix &lt;-<span class="st"> </span><span class="kw">matchMotifs</span>(motifs, counts_filtered,
                         <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19)

<span class="co"># computing deviations</span>
dev &lt;-<span class="st"> </span><span class="kw"><a href="reference/computeDeviations.html">computeDeviations</a></span>(<span class="dt">object =</span> counts_filtered, 
                                 <span class="dt">annotations =</span> motif_ix)</code></pre></div>
<p>See <a href="https://greenleaflab.github.io/chromVAR/">documentation website</a> for more information!</p>
</div>
</div>

  </div>

  <div class="col-md-3" id="sidebar">
    <h2>License</h2>
<p><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE">LICENSE</a></p>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Alicia Schep <br><small class="roles"> Author, maintainer </small> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/GreenleafLab/chromVAR"><img src="https://travis-ci.org/GreenleafLab/chromVAR.svg?branch=master" alt="Build Status"></a></li>
</ul>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Alicia Schep.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
