<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>chromVAR • chromVAR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">chromVAR</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/Articles/Counts.html">Counts</a>
    </li>
    <li>
      <a href="../articles/Articles/Annotations.html">Annotations</a>
    </li>
    <li>
      <a href="../articles/Articles/Deviations.html">Deviations</a>
    </li>
    <li>
      <a href="../articles/Articles/Applications.html">Applications</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="https://github.com/GreenleafLab/chromVAR">
    <span class="fa fa-github"></span>
     
    GitHub
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>chromVAR</h1>
                        <h4 class="author">Alicia Schep</h4>
            
          </div>

    
    
<div class="contents">
<p>chromVAR is an R package for the analysis of sparse chromatin accessibility. chromVAR takes as inputs aligned fragments (filtered for duplicates and low quality) from ATAC-seq or DNAse-seq experiments as well as genomic annotations such as motif positions. chromVAR computes for each annotation and each cell or sample a bias corrected “deviation” in accessibility from the expected accessibility based on the average of all the cells or samples.</p>
<p>This vignette covers basic usage of chromVAR with standard inputs. For more detailed documentation covering different options for various inputs and additional applications, view the additional vignettes on the <a href="https://greenleaflab.github.io/chromVAR/">documentation website</a>. For more description of the method and applications, please see the <a href="https://www.nature.com/nmeth/journal/vaop/ncurrent/full/nmeth.4401.html">publication</a> (<a href="http://greenleaf.stanford.edu/assets/pdf/nmeth.4401.pdf">pdf</a>, <a href="https://drive.google.com/file/d/0B8eUn6ZURmqvUjBCbE5Hc0p4UFU/view?usp=sharing">supplement</a>).</p>
<div id="loading-the-package" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-the-package" class="anchor"></a>Loading the package</h2>
<p>Use library or require to load the package and useful additional packages.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(chromVAR)
<span class="kw">library</span>(motifmatchr)
<span class="kw">library</span>(Matrix)
<span class="kw">library</span>(SummarizedExperiment)
<span class="kw">library</span>(BiocParallel)
<span class="kw">set.seed</span>(<span class="dv">2017</span>)</code></pre></div>
</div>
<div id="setting-multiprocessing-options" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-multiprocessing-options" class="anchor"></a>Setting multiprocessing options</h2>
<p>The package uses BiocParallel to do the multiprocessing. Check the documentation for BiocParallel to see available options. The settings can be set using the register function. For example, to use MulticoreParam with 8 cores:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/BiocParallel/topics/register">register</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/BiocParallel/topics/MulticoreParam-class">MulticoreParam</a></span>(<span class="dv">8</span>))</code></pre></div>
<p>To enable progress bars for multiprocessed tasks, use</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/BiocParallel/topics/register">register</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/BiocParallel/topics/MulticoreParam-class">MulticoreParam</a></span>(<span class="dv">8</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>))</code></pre></div>
<p>For Windows, <code>MulticoreParam</code> will not work, but you can use SnowParam:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/BiocParallel/topics/register">register</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/BiocParallel/topics/SnowParam-class">SnowParam</a></span>(<span class="dt">workers =</span> <span class="dv">1</span>, <span class="dt">type =</span> <span class="st">"SOCK"</span>))</code></pre></div>
<p>Even if you don’t want to use more than one core, it is best to explicitly register that choice using SerialParam:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/BiocParallel/topics/register">register</a></span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/BiocParallel/topics/SerialParam-class">SerialParam</a></span>())</code></pre></div>
<p>Please see the documentation for <a href="https://bioconductor.org/packages/release/bioc/html/BiocParallel.html">BiocParallel</a> for more information about the <code>register</code> function and the various options for multi-processing.</p>
</div>
<div id="reading-in-inputs" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-in-inputs" class="anchor"></a>Reading in inputs</h2>
<p>chromVAR takes as input a table of counts of fragments falling in open chromatin peaks. There are numerous software packages that enable the creation of count tables from epigenomics data; chromVAR also provides a method that is tailored for single-cell ATAC-seq counts. To learn more about the options for the count table and how to format a count table computed via other software, see the <a href="https://greenleaflab.github.io/chromVAR/articles/Articles/Counts.html">Counts section of the documentation website</a>.</p>
<div id="peaks" class="section level3">
<h3 class="hasAnchor">
<a href="#peaks" class="anchor"></a>Peaks</h3>
<p>For using chromVAR, it is recommended to use fixed-width, non-overlapping peaks. The method is fairly robust to the exact choice of peak width, but a width of 250 to 500 bp is recommended. See the <a href="https://drive.google.com/file/d/0B8eUn6ZURmqvUjBCbE5Hc0p4UFU/view?usp=sharing">supplement</a> of the paper for a disussion section and supplementary figures related to the choice of peak width.</p>
<p>If analyzing single cell data, it can make sense to use peaks derived from bulk ATAC or DNAse-seq data, either from the same population or a similar population (or possibly from a public resource, like the Roadmap Epigenomics project).</p>
<p>If combining multiple peak files from different populations, it is recommended to combine the peaks together. The <code>filterPeaks</code> function (demonstrated a bit further down this vignette) will reduce the peaks to a non-overlapping set based on which overlapping peak has stronger signal across all the data.</p>
<p>The function <code>getPeaks</code> reads in the peaks as a GenomicRanges object. We will show its use by reading in a tiny sample bed file. We’ll use the <code>sort_peaks</code> argument to indicate we want to sort the peaks.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">peakfile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/test_bed.txt"</span>, <span class="dt">package =</span> <span class="st">"chromVAR"</span>)
peaks &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getPeaks.html">getPeaks</a></span>(peakfile, <span class="dt">sort_peaks =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## Peaks sorted</code></pre>
<p>For reading in peaks in the narrowpeak format, chromVAR includes a function, <code>readNarrowpeaks</code>, that will read in the file, resize the peaks to a given size based on the <code>width</code> argument, and remove peaks that overlap a peak with stronger signal (if <code>non_overlapping</code> is set to TRUE – the default).</p>
</div>
<div id="counts" class="section level3">
<h3 class="hasAnchor">
<a href="#counts" class="anchor"></a>Counts</h3>
<p>The function <code>getCounts</code> returns a chromVARCounts object with a Matrix of fragment counts per sample/cell for each peak in assays. This data can be accessed with <code>counts(fragment_counts)</code>.The Matrix package is used so that if the matrix is sparse, the matrix will be stored as a sparse Matrix. We will demonstrate this function with a very tiny set of reads included in the package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bamfile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata/test_RG.bam"</span>, <span class="dt">package =</span> <span class="st">"chromVAR"</span>)
fragment_counts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getCounts.html">getCounts</a></span>(bamfile, 
                             peaks, 
                             <span class="dt">paired =</span>  <span class="ot">TRUE</span>, 
                             <span class="dt">by_rg =</span> <span class="ot">TRUE</span>, 
                             <span class="dt">format =</span> <span class="st">"bam"</span>, 
                             <span class="dt">colData =</span> <span class="kw">DataFrame</span>(<span class="dt">celltype =</span> <span class="st">"GM"</span>))</code></pre></div>
<pre><code>## Reading in file: /home/travis/R/Library/chromVAR/extdata/test_RG.bam</code></pre>
<pre><code>## Warning in doTryCatch(return(expr), name, parentenv, handler): length of
## NULL cannot be changed</code></pre>
<p>Here we passed only one bam file, but we can also pass a vector of bam file names. In that case, for the column data we would specify the appropriate value per file, e.g <code>DataFrame(celltype = c("GM","K562"))</code> if we were passing in one file for GM cells and one for K562 cells.</p>
<p>If RG tags are not used for combining multiple samples within a file, use <code>by_rg = FALSE</code>. For more on reading in counts, see the <a href="https://greenleaflab.github.io/chromVAR/articles/Articles/Counts.html">Counts section of the documentation website</a>.</p>
</div>
</div>
<div id="example-data" class="section level2">
<h2 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example data</h2>
<p>For the rest of the vignette, we will use a very small (but slightly larger than the previous example) data set of 10 GM cells and 10 H1 cells that has already been read in as a SummarizedExperiment object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(example_counts, <span class="dt">package =</span> <span class="st">"chromVAR"</span>)
<span class="kw">head</span>(example_counts)</code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 6 50 
## metadata(0):
## assays(1): counts
## rownames: NULL
## rowData names(3): score qval name
## colnames(50): singles-GM12878-140905-1 singles-GM12878-140905-2
##   ... singles-H1ESC-140820-24 singles-H1ESC-140820-25
## colData names(2): Cell_Type depth</code></pre>
</div>
<div id="getting-gc-content-of-peaks" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-gc-content-of-peaks" class="anchor"></a>Getting GC content of peaks</h2>
<p>The GC content will be used for determining background peaks. The function <code>addGCBias</code> returns an updated SummarizedExperiment with a new rowData column named “bias”. The function requires an input of a genome sequence, which can be provided as a BSgenome, FaFile, or DNAStringSet object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(BSgenome.Hsapiens.UCSC.hg19)
example_counts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/addGCBias.html">addGCBias</a></span>(example_counts, 
                            <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19)
<span class="kw">head</span>(<span class="kw"><a href="http://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class">rowData</a></span>(example_counts))</code></pre></div>
<pre><code>## DataFrame with 6 rows and 4 columns
##       score      qval        name      bias
##   &lt;integer&gt; &lt;numeric&gt; &lt;character&gt; &lt;numeric&gt;
## 1       259  25.99629   GM_peak_6     0.652
## 2        82   8.21494   H1_peak_7     0.680
## 3       156  15.65456  GM_peak_17     0.788
## 4        82   8.21494  H1_peak_13     0.674
## 5       140  14.03065  GM_peak_27     0.600
## 6       189  18.99529 GM_peak_29a     0.742</code></pre>
<p>Check out <code>available.genomes</code> from the BSgenome package for what genomes are available as BSgenome packages. For making your own BSgenome object, check out <code>BSgenomeForge</code>.</p>
</div>
<div id="filtering-inputs" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering-inputs" class="anchor"></a>Filtering inputs</h2>
<p>If working with single cell data, it is advisable to filter out samples with insufficient reads or a low proportion of reads in peaks as these may represent empty wells or dead cells. Two parameters are used for filtering – min_in_peaks and min_depth. If not provided (as above), these cutoffs are estimated based on the medians from the data. min_in_peaks is set to 0.5 times the median proportion of fragments in peaks. min_depth is set to the maximum of 500 or 10% of the median library size.</p>
<p>Unless <code>plot = FALSE</code> given as argument to function <code>filterSamples</code>, a plot will be generated.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#find indices of samples to keep</span>
counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="../reference/filterSamples.html">filterSamples</a></span>(example_counts, <span class="dt">min_depth =</span> <span class="dv">1500</span>, 
                                 <span class="dt">min_in_peaks =</span> <span class="fl">0.15</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>If shiny argument is set to TRUE (the default), a shiny gadget will pop up which allows you to play with the filtering parameters and see which cells pass filters or not.</p>
<p>To get just the plot of what is filtered, use <code>filterSamplesPlot</code>. By default, the plot is interactive– to set it as not interactive use <code>use_plotly = FALSE</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#find indices of samples to keep</span>
filtering_plot &lt;-<span class="st"> </span><span class="kw"><a href="../reference/filterSamplesPlot.html">filterSamplesPlot</a></span>(example_counts, <span class="dt">min_depth =</span> <span class="dv">1500</span>, 
                                    <span class="dt">min_in_peaks =</span> <span class="fl">0.15</span>, <span class="dt">use_plotly =</span> <span class="ot">FALSE</span>)
filtering_plot</code></pre></div>
<p><img src="Introduction_files/figure-html/filter_plot-1.png" width="576"></p>
<p>To instead return the indexes of the samples to keep instead of a new SummarizedExperiment object, use ix_return = TRUE.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ix &lt;-<span class="st"> </span><span class="kw"><a href="../reference/filterSamples.html">filterSamples</a></span>(example_counts, <span class="dt">ix_return =</span> <span class="ot">TRUE</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre><code>## min_in_peaks set to 0.105</code></pre>
<pre><code>## min_depth set to 1120.65</code></pre>
<p>For both bulk and single cell data, peaks should be filtered based on having at least a certain number of fragments. At minimum, each peak should have at least one fragment across all the samples (it might be possible to have peaks with zero reads due to using a peak set defined by other data). Otherwise, downstream functions won’t work. The function <code>filterPeaks</code> will also reduce the peak set to non-overlapping peaks (keeping the peak with higher counts for peaks that overlap) if non_overlapping argument is set to TRUE (which is the default).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts_filtered &lt;-<span class="st"> </span><span class="kw"><a href="../reference/filterPeaks.html">filterPeaks</a></span>(counts_filtered, <span class="dt">non_overlapping =</span> <span class="ot">TRUE</span>)</code></pre></div>
</div>
<div id="get-motifs-and-what-peaks-contain-motifs" class="section level2">
<h2 class="hasAnchor">
<a href="#get-motifs-and-what-peaks-contain-motifs" class="anchor"></a>Get motifs and what peaks contain motifs</h2>
<p>The function <code>getJasparMotifs</code> fetches motifs from the JASPAR database.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">motifs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getJasparMotifs.html">getJasparMotifs</a></span>()</code></pre></div>
<p>The function getJasparMotifs() by default gets human motifs from JASPAR core database. For other species motifs, change the species argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">yeast_motifs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getJasparMotifs.html">getJasparMotifs</a></span>(<span class="dt">species =</span> <span class="st">"Saccharomyces cerevisiae"</span>)</code></pre></div>
<p>For using a collection other than core, use the <code>collection</code> argument. Options include: “CORE”, “CNE”, “PHYLOFACTS”, “SPLICE”, “POLII”, “FAM”, “PBM”, “PBM_HOMEO”, “PBM_HLH”.</p>
<p>The <code>getJasparMotifs</code> function is simply a wrapper around <code>getMatrixSet</code> from TFBSTools– you can also use that function to fetch motifs from JASPAR if you prefer, and/or check out the documentation for that function for more information.</p>
<p>The function <code>matchMotifs</code> from the motifmatchr package finds which peaks contain which motifs. By default, it returns a SummarizedExperiment object, which contains a sparse matrix indicating motif match or not.The function requires an input of a genome sequence, which can be provided as a BSgenome, FaFile, or DNAStringSet object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(motifmatchr)
motif_ix &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/motifmatchr/topics/matchMotifs">matchMotifs</a></span>(motifs, counts_filtered, 
                        <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19)</code></pre></div>
<p>One option is the p.cutoff for determing how stringent motif calling should be. The default value is 0.00005, which tends to give reasonable numbers of motif matches for human motifs.</p>
<p>Instead of returning just motif matches, the function can also return additional matrices (stored as assays) with the number of motif matches per peak and the maximum motif score per peak. For this additional information, use <code>out = scores</code>. To return the actual positions of motif matches, use <code>out = positions</code>. Either the output with <code>out = matches</code> or <code>out = scores</code> can be passed to the computeDeviations function.</p>
<p>If instead of using known motifs, you want to use all kmers of a certain length, the <code>matchKmers</code> function can be used. For more about using kmers as inputs, see the the <a href="https://greenleaflab.github.io/chromVAR/articles/Articles/Counts.html">Annotations section of the documentation website</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">kmer_ix &lt;-<span class="st"> </span><span class="kw"><a href="../reference/matchKmers.html">matchKmers</a></span>(<span class="dv">6</span>, counts_filtered, 
                      <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19)</code></pre></div>
</div>
<div id="compute-deviations" class="section level2">
<h2 class="hasAnchor">
<a href="#compute-deviations" class="anchor"></a>Compute deviations</h2>
<p>The function <code>computeDeviations</code> returns a SummarizedExperiment with two “assays”. The first matrix (accessible via <code><a href="../reference/deviations.html">deviations(dev)</a></code> or <code>assays(dev)$deviations</code>) will give the bias corrected “deviation” in accessibility for each set of peaks (rows) for each cell or sample (columns). This metric represent how accessible the set of peaks is relative to the expectation based on equal chromatin accessibility profiles across cells/samples, normalized by a set of background peak sets matched for GC and average accessability. The second matrix (<code><a href="../reference/deviationScores.html">deviationScores(dev)</a></code> or <code>assays(deviations)$z</code>) gives the deviation Z-score, which takes into account how likely such a score would occur if randomly sampling sets of beaks with similar GC content and average accessibility.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dev &lt;-<span class="st"> </span><span class="kw"><a href="../reference/computeDeviations.html">computeDeviations</a></span>(<span class="dt">object =</span> counts_filtered, <span class="dt">annotations =</span> motif_ix)</code></pre></div>
</div>
<div id="background-peaks" class="section level2">
<h2 class="hasAnchor">
<a href="#background-peaks" class="anchor"></a>Background Peaks</h2>
<p>The function computeDeviations will use a set of background peaks for normalizing the deviation scores. This computation is done internally by default and not returned – to have greater control over this step, a user can run the <code>getBackgroundPeaks</code> function themselves and pass the result to computeDeviations under the background_peaks parameter.</p>
<p>Background peaks are peaks that are similar to a peak in GC content and average accessibility.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getBackgroundPeaks.html">getBackgroundPeaks</a></span>(<span class="dt">object =</span> counts_filtered)</code></pre></div>
<p>The result from <code>getBackgroundPeaks</code> is a matrix of indices, where each column represents the index of the peak that is a background peak.</p>
<p>To use the background peaks computed, simply add those to the call to computeDeviations:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dev &lt;-<span class="st"> </span><span class="kw"><a href="../reference/computeDeviations.html">computeDeviations</a></span>(<span class="dt">object =</span> counts_filtered, <span class="dt">annotations =</span> motif_ix,
                         <span class="dt">background_peaks =</span> bg)</code></pre></div>
</div>
<div id="variability" class="section level2">
<h2 class="hasAnchor">
<a href="#variability" class="anchor"></a>Variability</h2>
<p>The function <code>computeVariability</code> returns a data.frame that contains the variability (standard deviation of the z scores computed above across all cell/samples for a set of peaks), bootstrap confidence intervals for that variability (by resampling cells/samples), and a p-value for the variability being greater than the null hypothesis of 1.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">variability &lt;-<span class="st"> </span><span class="kw"><a href="../reference/computeVariability.html">computeVariability</a></span>(dev)

<span class="kw"><a href="../reference/plotVariability.html">plotVariability</a></span>(variability, <span class="dt">use_plotly =</span> <span class="ot">FALSE</span>) </code></pre></div>
<p><img src="Introduction_files/figure-html/variability-1.png" width="576"></p>
<p><code>plotVariability</code> takes the output of <code>computeVariability</code> and returns a plot of rank sorted annotation sets and their variability. By default, the plot will be interactive, unless you set <code>use_plotly = FALSE</code>.</p>
</div>
<div id="visualizing-deviations" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizing-deviations" class="anchor"></a>Visualizing Deviations</h2>
<p>For visualizing cells, it can be useful to project the deviation values into two dimension using TSNE. A convenience function for doing so is provided in <code>deviationsTsne</code>. If running in an interactive session, shiny can be set to TRUE to load up a shiny gadget for exploring parameters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tsne_results &lt;-<span class="st"> </span><span class="kw"><a href="../reference/deviationsTsne.html">deviationsTsne</a></span>(dev, <span class="dt">threshold =</span> <span class="fl">1.5</span>, <span class="dt">perplexity =</span> <span class="dv">10</span>)</code></pre></div>
<p>To plot the results, <code>plotDeviationsTsne</code> can be used. If running in an interactive session or an interactive Rmarkdown document, shiny can be set to TRUE to generate a shiny widget. Here we will show static results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tsne_plots &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotDeviationsTsne.html">plotDeviationsTsne</a></span>(dev, tsne_results, 
                                 <span class="dt">annotation_name =</span> <span class="st">"TEAD3"</span>, 
                                   <span class="dt">sample_column =</span> <span class="st">"Cell_Type"</span>, 
                                 <span class="dt">shiny =</span> <span class="ot">FALSE</span>)
tsne_plots[[<span class="dv">1</span>]]</code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-21-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tsne_plots[[<span class="dv">2</span>]]</code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-21-2.png" width="672"></p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">Sys.Date</span>()</code></pre></div>
<pre><code>## [1] "2017-11-10"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R Under development (unstable) (2017-11-10 r73707)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.5 LTS
## 
## Matrix products: default
## BLAS: /home/travis/R-bin/lib/R/lib/libRblas.so
## LAPACK: /home/travis/R-bin/lib/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    methods   stats     graphics  grDevices utils    
## [8] datasets  base     
## 
## other attached packages:
##  [1] BSgenome.Hsapiens.UCSC.hg19_1.4.0 BSgenome_1.46.0                  
##  [3] rtracklayer_1.38.0                Biostrings_2.46.0                
##  [5] XVector_0.18.0                    BiocParallel_1.12.0              
##  [7] SummarizedExperiment_1.8.0        DelayedArray_0.4.1               
##  [9] matrixStats_0.52.2                Biobase_2.38.0                   
## [11] GenomicRanges_1.30.0              GenomeInfoDb_1.14.0              
## [13] IRanges_2.12.0                    S4Vectors_0.16.0                 
## [15] BiocGenerics_0.24.0               Matrix_1.2-11                    
## [17] motifmatchr_1.1.1                 chromVAR_1.1.1                   
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                DirichletMultinomial_1.20.0
##  [3] TFBSTools_1.16.0            bit64_0.9-7                
##  [5] RColorBrewer_1.1-2          httr_1.3.1                 
##  [7] rprojroot_1.2               tools_3.5.0                
##  [9] backports_1.1.1             R6_2.2.2                   
## [11] DT_0.2                      seqLogo_1.44.0             
## [13] DBI_0.7                     lazyeval_0.2.0             
## [15] colorspace_1.3-2            bit_1.1-12                 
## [17] compiler_3.5.0              plotly_4.7.1               
## [19] labeling_0.3                caTools_1.17.1             
## [21] scales_0.5.0                readr_1.1.1                
## [23] nabor_0.4.7                 stringr_1.2.0              
## [25] digest_0.6.12               Rsamtools_1.30.0           
## [27] rmarkdown_1.7               R.utils_2.6.0              
## [29] pkgconfig_2.0.1             htmltools_0.3.6            
## [31] htmlwidgets_0.9             rlang_0.1.2.9000           
## [33] RSQLite_2.0                 VGAM_1.0-4                 
## [35] shiny_1.0.5                 bindr_0.1                  
## [37] jsonlite_1.5                gtools_3.5.0               
## [39] dplyr_0.7.3                 R.oo_1.21.0                
## [41] RCurl_1.95-4.8              magrittr_1.5               
## [43] GO.db_3.5.0                 GenomeInfoDbData_0.99.1    
## [45] Rcpp_0.12.13                munsell_0.4.3              
## [47] R.methodsS3_1.7.1           stringi_1.1.5              
## [49] yaml_2.1.14                 zlibbioc_1.24.0            
## [51] Rtsne_0.13                  plyr_1.8.4                 
## [53] grid_3.5.0                  blob_1.1.0                 
## [55] miniUI_0.1.1                CNEr_1.14.0                
## [57] lattice_0.20-35             splines_3.5.0              
## [59] annotate_1.56.0             hms_0.3                    
## [61] KEGGREST_1.18.0             knitr_1.17                 
## [63] JASPAR2016_1.6.0            codetools_0.2-15           
## [65] reshape2_1.4.2              TFMPvalue_0.0.6            
## [67] XML_3.98-1.9                glue_1.1.1                 
## [69] evaluate_0.10.1             data.table_1.10.4          
## [71] png_0.1-7                   httpuv_1.3.5               
## [73] tidyr_0.7.1                 purrr_0.2.3                
## [75] gtable_0.2.0                poweRlaw_0.70.1            
## [77] assertthat_0.2.0            ggplot2_2.2.1              
## [79] mime_0.5                    xtable_1.8-2               
## [81] viridisLite_0.2.0           tibble_1.3.4               
## [83] GenomicAlignments_1.14.0    AnnotationDbi_1.40.0       
## [85] memoise_1.1.0               bindrcpp_0.2</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#loading-the-package">Loading the package</a></li>
      <li><a href="#setting-multiprocessing-options">Setting multiprocessing options</a></li>
      <li><a href="#reading-in-inputs">Reading in inputs</a></li>
      <li><a href="#example-data">Example data</a></li>
      <li><a href="#getting-gc-content-of-peaks">Getting GC content of peaks</a></li>
      <li><a href="#filtering-inputs">Filtering inputs</a></li>
      <li><a href="#get-motifs-and-what-peaks-contain-motifs">Get motifs and what peaks contain motifs</a></li>
      <li><a href="#compute-deviations">Compute deviations</a></li>
      <li><a href="#background-peaks">Background Peaks</a></li>
      <li><a href="#variability">Variability</a></li>
      <li><a href="#visualizing-deviations">Visualizing Deviations</a></li>
      <li><a href="#session-info">Session Info</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Alicia Schep.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
